### ISSUE TITLE
mPMT integration

### ISSUE THREAD

**Issue by JostMigenda** [HKDAQ/FLOWER#12]
As far as I know, we don’t currently have an energy reconstruction that supports mPMTs. Over the next couple of months, we’ll therefore need to integrate mPMTs in FLOWER.

Broadly speaking, this will require two types of changes to FLOWER: changes to the energy reconstruction itself and “bookkeeping changes” (e.g. if WCSim stores mPMT hits slightly differently from 50cm PMT hits).

I’ll discuss the first type in a follow-up comment on this issue. I’ll leave the second type of changes to someone who is familiar with the mPMT implementation in WCSim.

**Comment by JostMigenda** [comment]
I spent some time thinking about how to adapt the FLOWER algorithm (in [WCSimFLOWER::GetEnergy](https://github.com/HKDAQ/FLOWER/blob/master/WCSimFLOWER.cpp#L152)) to support mPMTs. Here are the main changes that are required:

**CorrectHitTimes():** shouldn’t depend on PMT type

**FindMaxTimeInterval():** shouldn’t depend on PMT type

**GetNEff():** We’ll need to make a few changes here.

* [Occupancy correction](https://github.com/HKDAQ/FLOWER/blob/master/WCSimFLOWER.cpp#L346): This will almost certainly not be necessary for mPMTs, since the 7.5cm PMTs in the mPMTs have a ~40× smaller surface than the 50cm PMTs (which are already unlikely to receive more than one hit). However, if we do want to implement it, it would probably be best to consider hits on other PMTs within the same mPMT module; not hits in neighbouring mPMTs or 50cm PMTs.

* [Delayed Hits correction](https://github.com/HKDAQ/FLOWER/blob/master/WCSimFLOWER.cpp#L369): This should in principle work the same, but since it depends on dark noise rate we’ll need to treat both types of PMT separately.

* [Dark Noise correction](https://github.com/HKDAQ/FLOWER/blob/master/WCSimFLOWER.cpp#L372): Again, this should in principle work the same, but we’ll need to treat both types of PMT separately since they have different dark noise rates.

* [Effective Photocoverage correction](https://github.com/HKDAQ/FLOWER/blob/master/WCSimFLOWER.cpp#L375): This is tricky. As an initial, very rough stopgap measure, we can probably get away with multiplying the effective coverages in `fEffCoverages` by a constant factor (e.g. 1.1 if the total photosensitive area of mPMTs is 10% of that of 50cm PMTs). That should be very accurate for events in the middle of the detector (where the angle between the incoming photon and ID wall is close to 90°, i.e. theta≈0), but will lose accuracy for events closer to the wall. Since the geometry of the mPMTs is more complicated, I’m not sure whether a reasonable toy model can be used to estimate the effective coverage at higher theta. To determine this correction precisely, we would have to run a full MC with sufficiently high photon statistics.

**CorrectEnergy():** We’ll need to calibrate this Neff-Energy relation new for every detector configuration. (Run detector simulation of a few thousand monoenergetic electrons at various energies, determine average Neff for each energy, then perform a fit.)


In summary, I think the changes required to adapt the energy reconstruction itself are not too large; recalibrating the Neff-Energy relation will probably be the most time consuming.
We can think about more refinements later, of course (for example, if an event is reconstructed to be outside the field of view of a 7.5cm PMTs, it could be excluded from the energy reconstruction to eliminate some dark noise hits) but I suspect that the minimal changes listed here would already give a good result.

**Comment by JostMigenda** [comment]
Got most of these changes implemented in the PR. A few notes:
* mPMT hits don’t receive occupancy correction currently, as discussed above. (However, when calculating the occupancy correction for 50cm PMTs, hits on nearby mPMTs _are_ counted.)
* The effective photocoverage correction is simply scaled with the photosensitive area at the moment.
* Neff-to-Energy calibration is still TBD.
